File: geometry-viewer/Matrix.js
class Matrix {
    /** @private @type {number} */
    rows;
    /** @private @type {number} */
    columns;
    /** @private @type {number[][]} */
    data;
    /**
     *  @description create a new matrix
     *  @param {number} rows Number of rows
     *  @param {number} columns Number of columns
    */
    constructor(rows, columns) {
        this.rows = rows;
        this.columns = columns;
        this.data = Array.from({ length: rows }, () => Array(columns).fill(0));
    }
    getColumnsSize() { return this.columns; }
    getRowsSize()    { return this.rows; }
    /**
     * @description set a value in the matrix
     * @param {number} rows Number of rows
     * @param {number} columns Number of columns
     * @param {any} value Value to set
     */
    set(row, column, value) {
        if (! this.isInRange(row, column)) throw new Error("out of range");
        this.data[row][column] = value;
    }
    /**
     * @description set a row in the matrix
     * @param {number} row Number of rows
     * @param {any[]} values Values to set in the row
     */
    setRow(row, values) {
        if (! this.isInRange(row, 0)) throw new Error("out of range");
        if (values.length !== this.columns) throw new Error("invalid length");
        this.data[row][i] = [...values];
    }
    setColumn(column, values) {
        if (! this.isInRange(0, column)) throw new Error("out of range");
        if (values.length !== this.rows) throw new Error("invalid length");
        for (let i = 0; i < this.rows; i++) {
            this.data[i][column] = values[i];
        }
    }
    /**
     * @description set all values in the matrix
     * @param {any[][]} values Values to set in the matrix
    */
    setAll(values) {
        if (values.length !== this.rows || values[0].length !== this.columns) throw new Error("invalid length");
        for (let i = 0; i < this.rows; i++)
        for (let j = 0; j < this.columns; j++) {
            this.data[i][j] = values[i][j];
        }
    }
    /**
     * @description get a value in the matrix
     * @param {number} rows Number of rows
     * @param {number} columns Number of columns
     * @returns {any}
     */
    get(row, column) {
        if (! this.isInRange(row, column)) throw new Error("out of range");
        return this.data[row][column];
    }
    /**
     * @description get a row in the matrix
     * @param {number} row Number of rows
     * @returns {any[]}
     */
    getRow(row) {
        if (! this.isInRange(row, 0)) throw new Error("out of range");
        return [...this.data[row]];
    }
    /**
     * @description get a column in the matrix
     * @param {number} column Number of columns
     * @returns {any[]}
     */
    getColumn(column) {
        if (! this.isInRange(0, column)) throw new Error("out of range");
        return this.data.map(row => row[column]);;
    }
    /**
     * @protected
     * @description check if a value is in the matrix
     * @param {number} rows Number of rows
     * @param {number} columns Number of columns
     * @returns {boolean}
     */
    isInRange(row, column) {
        return (
            row >= 0 &&
            row < this.rows &&
            column >= 0 &&
            column < this.columns
        );
    }
    /**
     * @description iterate over the matrix
     * @param {(value: any, row: number, column: number) => void} callback Callback function
    */
    forEach(callback) {
        for (let i = 0; i < this.rows; i++)
        for (let j = 0; j < this.columns; j++) {
            callback(this.get(i, j), i, j);
        }
    }
    /**
     * @description clone this matrix
     * @returns {Matrix}
     */
    clone() {
        const clone = new Matrix(this.rows, this.columns);
        for (let i = 0; i < this.rows; i++)
        for (let j = 0; j < this.columns; j++) {
            clone.set(i, j, this.get(i, j));
        }
        return clone;
    }
    static Multiply(matrix1, matrix2) {
        if (matrix1.getColumnsSize() !== matrix2.getRowsSize()) throw new Error("invalid dimensions");
        const result = new Matrix(matrix1.getRowsSize(), matrix2.getColumnsSize());
        for (let filaCount = 0; filaCount < result.getRowsSize(); filaCount++)
        for (let ColumnCount = 0; ColumnCount < result.getColumnsSize(); ColumnCount++) {
            let sum = 0;
            for (let k = 0; k < matrix1.getColumnsSize(); k++) {
                sum += matrix1.get(filaCount, k) * matrix2.get(k, ColumnCount);
            }
            result.set(filaCount, ColumnCount, sum);
        }
        return result;
    }
    /**
     * @description clone a matrix
     * @param {Matrix | null} matrix Matrix to clone
     * @returns {Matrix}
     */
    static clone(matrix) {
        const clone = new Matrix(this.rows, this.columns);
        for (let i = 0; i < this.rows; i++)
        for (let j = 0; j < this.columns; j++) {
            clone.set(i, j, matrix.get(i, j));
        }
        return clone;
    }
    toString() {
        return this.data.map(row => `|${row.join(", ")}|`).join("\n");
    }
}
export default Matrix;

File: src/main/java/com/mysaml/mc/tp/geometry/Matrix.java
package com.mysaml.mc.tp.geometry;

import java.util.Arrays;

public class Matrix {
    private int rows;
    private int columns;
    private double[][] data;

    /**
     * Creates a new matrix
     * @param rows Number of rows
     * @param columns Number of columns
     */
    public Matrix(int rows, int columns) {
        this.rows = rows;
        this.columns = columns;
        this.data = new double[rows][columns];
        for (int i = 0; i < rows; i++) {
            Arrays.fill(this.data[i], 0);
        }
    }

    public int getColumnsSize() { return this.columns; }
    public int getRowsSize() { return this.rows; }

    /**
     * Sets a value in the matrix
     * @param row Row index
     * @param column Column index
     * @param value Value to set
     */
    public void set(int row, int column, double value) {
        if (!this.isInRange(row, column)) throw new IndexOutOfBoundsException("out of range");
        this.data[row][column] = value;
    }

    /**
     * Sets a row in the matrix
     * @param row Row index
     * @param values Values to set in the row
     */
    public void setRow(int row, double[] values) {
        if (!this.isInRange(row, 0)) throw new IndexOutOfBoundsException("out of range");
        if (values.length != this.columns) throw new IllegalArgumentException("invalid length");
        this.data[row] = Arrays.copyOf(values, values.length);
    }

    /**
     * Sets a column in the matrix
     * @param column Column index
     * @param values Values to set in the column
     */
    public void setColumn(int column, double[] values) {
        if (!this.isInRange(0, column)) throw new IndexOutOfBoundsException("out of range");
        if (values.length != this.rows) throw new IllegalArgumentException("invalid length");
        for (int i = 0; i < this.rows; i++) {
            this.data[i][column] = values[i];
        }
    }

    /**
     * Sets all values in the matrix
     * @param values Values to set in the matrix
     */
    public void setAll(double[][] values) {
        if (values.length != this.rows || values[0].length != this.columns) throw new IllegalArgumentException("invalid length");
        for (int i = 0; i < this.rows; i++) {
            System.arraycopy(values[i], 0, this.data[i], 0, this.columns);
        }
    }

    /**
     * Gets a value in the matrix
     * @param row Row index
     * @param column Column index
     * @return Value at the specified position
     */
    public double get(int row, int column) {
        if (!this.isInRange(row, column)) throw new IndexOutOfBoundsException("out of range");
        return this.data[row][column];
    }

    /**
     * Gets a row in the matrix
     * @param row Row index
     * @return Values in the specified row
     */
    public double[] getRow(int row) {
        if (!this.isInRange(row, 0)) throw new IndexOutOfBoundsException("out of range");
        return Arrays.copyOf(this.data[row], this.columns);
    }

    /**
     * Gets a column in the matrix
     * @param column Column index
     * @return Values in the specified column
     */
    public double[] getColumn(int column) {
        if (!this.isInRange(0, column)) throw new IndexOutOfBoundsException("out of range");
        double[] columnData = new double[this.rows];
        for (int i = 0; i < this.rows; i++) {
            columnData[i] = this.data[i][column];
        }
        return columnData;
    }

    /**
     * Checks if a value is in the matrix
     * @param row Row index
     * @param column Column index
     * @return True if the value is in range, otherwise false
     */
    protected boolean isInRange(int row, int column) {
        return (
            row >= 0 &&
            row < this.rows &&
            column >= 0 &&
            column < this.columns
        );
    }

    /**
     * Iterates over the matrix
     * @param callback Callback function
     */
    public void forEach(MatrixCallback callback) {
        for (int i = 0; i < this.rows; i++)
        for (int j = 0; j < this.columns; j++) {
            callback.apply(this.get(i, j), i, j);
        }
    }

    /**
     * Clones this matrix
     * @return A clone of this matrix
     */
    public Matrix clone() {
        Matrix clone = new Matrix(this.rows, this.columns);
        for (int i = 0; i < this.rows; i++)
        for (int j = 0; j < this.columns; j++) {
            clone.set(i, j, this.get(i, j));
        }
        return clone;
    }

    /**
     * Multiplies two matrices
     * @param matrix1 The first matrix
     * @param matrix2 The second matrix
     * @return The result of the multiplication
     */
    public static Matrix multiply(Matrix matrix1, Matrix matrix2) {
        if (matrix1.getColumnsSize() != matrix2.getRowsSize()) throw new IllegalArgumentException("invalid dimensions");
        Matrix result = new Matrix(matrix1.getRowsSize(), matrix2.getColumnsSize());
        for (int i = 0; i < result.getRowsSize(); i++)
        for (int j = 0; j < result.getColumnsSize(); j++) {
            double sum = 0;
            for (int k = 0; k < matrix1.getColumnsSize(); k++) {
                sum += matrix1.get(i, k) * matrix2.get(k, j);
            }
            result.set(i, j, sum);
        }
        return result;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < this.rows; i++) {
            builder.append("|");
            for (int j = 0; j < this.columns; j++) {
                builder.append(this.data[i][j]);
                if (j < this.columns - 1) {
                    builder.append(", ");
                }
            }
            builder.append("|\n");
        }
        return builder.toString();
    }

    @FunctionalInterface
    public interface MatrixCallback {
        void apply(double value, int row, int column);
    }
}